src/cart/crt_group.c

```diff
@@ -316,19 +321,86 @@ grp_li_uri_set(struct crt_lookup_item *li, int tag, const char *uri)
 		ui->ui_ref = 0;
 		ui->ui_initialized = 1;
 		ui->ui_rank = li->li_rank;
-		D_STRNDUP(ui->ui_uri[tag], uri, CRT_ADDR_STR_MAX_LEN);
-		if (!ui->ui_uri[tag]) {
-			D_FREE_PTR(ui);
-			D_GOTO(exit, rc = -DER_NOMEM);
+
+		if (crt_provider_is_contig_ep(crt_gdata.cg_na_plugin)) {
+			D_STRNDUP(tmp_uri, uri, CRT_ADDR_STR_MAX_LEN);
+			if (!tmp_uri)
+				D_GOTO(exit, rc = -DER_NOMEM);
+
+			/* For now we assume contiguous endpoint providers are
+			 * port based. Based on that we generate URIs for every
+			 * tag of the rank from the base port.
+			 *
+			 * Port-based providers have form of
+			 * string:port
+			 *
+			 * Parse both parts out
+			 */
+			p = tmp_uri;
+			while (*p != '\0')
+				p++;
+			while (*p != ':' && p != tmp_uri)
+				p--;
+
+			if (p == tmp_uri) {
+				D_ERROR("Badly formed URI '%s'\n", tmp_uri);
+				D_FREE(tmp_uri);
+				D_GOTO(exit, rc = -DER_INVAL);
+			}
+
+			/* Split <string> from <port> part in URI */
+			*p = '\0';
+			p++;
+			base_port = atoi(p) - tag;
+
+			if (base_port <= 0) {
+				D_ERROR("Failed to parse uri=%s correctly\n",
+					tmp_uri);
+				D_FREE(tmp_uri);
+				D_GOTO(exit, rc = -DER_INVAL);
+			}
+
+			for (i = 0; i < 255; i++) {
```

Comment:
255?   Maybe define a constant?

Reply:
Note: strrchr may not be supported by all C Libraries.  I don't know when it was put into the standards but I came across this problem at my previous place,

Reply:
Still need to use macro name for string size.